src/sources/util/tcp.rs

```diff
@@ -168,6 +168,15 @@ where
                         let peer_addr = socket.peer_addr();
                         let span = info_span!("connection", %peer_addr);
 
+                        // async block if we are over the connection limit
+                        let permit = match connection_semaphore {
+                            None => None,
+                            Some(semaphore) => {
+                                // is the semaphore is closed, or there is no limit, this will return `None`
+                                semaphore.acquire_owned().await.ok()
+                            }
+                        };
```

Comment:
By checking the semaphore at this point, we'll potentially block after accepting a new connection rather than before. It's possible that it doesn't really matter, but it seems like there could be some behavior downsides. From the perspective of a load balancer, if we accept a connection but don't actually begin to process it then we lose the possibility of retrying that connection attempt against a less busy node. Especially since these TCP connections can be relatively long-lived.

Reply:
I moved this to before the connection is accepted.